nose2/suite.py

```diff
@@ -65,27 +73,43 @@ def setUpTest(self, test):
             return
         if getattr(test, '_layer_wasSetUp', False):
             return
+
+        event = events.StartLayerSetupTestEvent(self.layer, test)
+        self.session.hooks.startLayerSetupTest(event)
+
         self._allLayers(test, 'testSetUp')
         test._layer_wasSetUp = True
 
+        event = events.StopLayerSetupTestEvent(self.layer, test)
+        self.session.hooks.stopLayerSetupTest(event)
+
     def tearDownTest(self, test):
-        # FIXME hook call
         if self.layer is None:
             return
         if not getattr(test, '_layer_wasSetUp', None):
             return
+
+        event = events.StartLayerTeardownTestEvent(self.layer, test)
+        self.session.hooks.startLayerTeardownTest(event)
+
         self._allLayers(test, 'testTearDown', reverse=True)
+
+        event = events.StopLayerTeardownTestEvent(self.layer, test)
+        self.session.hooks.stopLayerTeardownTest(event)
         delattr(test, '_layer_wasSetUp')
 
     def tearDown(self):
-        # FIXME hook call
         if self.layer is None:
             return
 
         teardown = self._getBoundClassmethod(self.layer, 'tearDown')
+        event = events.StartLayerTeardownEvent(self.layer)
+        self.session.hooks.startLayerTeardown(event)
         if teardown:
             teardown()
             log.debug('tearDown layer %s called', self.layer)
+        event = events.StopLayerTeardownEvent(self.layer)
```

Comment:
With this implementation, the hook is called even if no teardown is defined in the layer. I noticed that because I was looking for a way to know when I'm entering/exiting a layer, and this hooks provides a very good way to tell when you're exiting a layer :)

Some comment for the `startLayerSetup` hook.